### PR TITLE
Workflow Authoring: initial methods

### PR DESCRIPTION
**Pull request by tmacam**
# Description

(This PR is a re-submission of PR dapr/dotnet-sdk#973 with extra fixes when appropriate)

This is a work in progress PR for dapr workflow initial methods for dotnet-sdk.
This needs e2e tests still.

## Issue reference

We strive to have all PR being opened based on an issue, where the problem or feature have been discussed prior to implementation.

Please reference the issue this PR will close: #979

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [X] Code compiles correctly
* [ ] Created/updated tests
* [ ] Extended the documentation

### CODE REVIEW

**Comment by cgillum** [comment]
Hi @tmacam, can you please accept my PR [here](https://github.com/tmacam/dapr-dotnet-sdk/pull/1), which will update the Microsoft.DurableTask.* dependencies in this PR to v1.0.0-rc.1? There were several breaking changes and important bug fixes in that release, and we should make sure that those changes are reflected in this new Dapr Workflow SDK.

File: src/Dapr.Workflow/WorkflowServiceCollectionExtensions.cs
Comment (philliphoff) [inline src/Dapr.Workflow/WorkflowServiceCollectionExtensions.cs:31]: `AddDaprWorkflow()`?

File: src/Dapr.Workflow/WorkflowMetadata.cs
Comment (philliphoff) [inline src/Dapr.Workflow/WorkflowMetadata.cs:23]: Nit: `details` instead of `metadata`?

File: examples/Workflow/WorkflowWebApp/Properties/launchSettings.json
Comment (philliphoff) [inline examples/Workflow/WorkflowWebApp/Properties/launchSettings.json:1]: Should this be in the repo? (I was asked to remove it from samples in another repo.)

File: examples/Workflow/WorkflowWebApp/WorkflowWebApp.csproj
Comment (philliphoff) [inline examples/Workflow/WorkflowWebApp/WorkflowWebApp.csproj:11]: `10.0`?  (As .NET 6 is compatible only with C# <=10.)

File: examples/Workflow/WorkflowWebApp/Program.cs
Comment (philliphoff) [inline examples/Workflow/WorkflowWebApp/Program.cs:50]: Nit: perhaps choose a default other than `8080`, as that could conflict with Dapr's own dashboard?

File: src/Dapr.Workflow/Dapr.Workflow.csproj
Comment (philliphoff) [inline src/Dapr.Workflow/Dapr.Workflow.csproj:17]: Nit: `1.9.0`?

Comment (tmacam) [reply src/Dapr.Workflow/Dapr.Workflow.csproj:17]: Yes, support for pluggable components only appears on 1.9+

File: examples/Workflow/WorkflowWebApp/Program.cs
Comment (tmacam) [reply examples/Workflow/WorkflowWebApp/Program.cs:50]: Not sure about dapr dashboard but, for instance, examples and [quickstarts](https://docs.dapr.io/getting-started/quickstarts/) tend to stay away from this port. Might be worth using a different one just to be on the safe side.

File: examples/Workflow/WorkflowWebApp/WorkflowWebApp.csproj
Comment (tmacam) [reply examples/Workflow/WorkflowWebApp/WorkflowWebApp.csproj:11]: Your call. As I understand we have a bit more freedom on choosing newer SDKs/language versions for pluggable components than on the regular SDKs

File: src/Dapr.Workflow/WorkflowServiceCollectionExtensions.cs
Comment (cgillum) [reply src/Dapr.Workflow/WorkflowServiceCollectionExtensions.cs:31]: I could go either way, but the Actors SDK names the equivalent extension method `AddActors()` (no "Dapr") so `AddDaprWorkflow` would be a bit inconsistent with the actor SDK.

File: src/Dapr.Workflow/WorkflowContext.cs
Comment (philliphoff) [inline src/Dapr.Workflow/WorkflowContext.cs:67]: Does this inner context not have an overload that accepts a cancellation token?  (I expect not, given the timeout, but just checking.)

File: examples/Workflow/WorkflowWebApp/Program.cs
Comment (philliphoff) [inline examples/Workflow/WorkflowWebApp/Program.cs:12]: Nit: remove extra space and use `String.Empty`?

File: src/Dapr.Workflow/WorkflowContext.cs
Comment (philliphoff) [inline src/Dapr.Workflow/WorkflowContext.cs:75]: Similar question re: support for cancellation.

**Review by philliphoff — COMMENTED** [review]
Just some nits and comments re: naming, overall looks good though.

File: src/Dapr.Workflow/WorkflowClient.cs
Comment (RyanLettieri) [inline src/Dapr.Workflow/WorkflowClient.cs:42]: Should we also have the param descriptions in this method header similar to how we have done above?

File: src/Dapr.Workflow/WorkflowRuntimeOptions.cs
Comment (RyanLettieri) [inline src/Dapr.Workflow/WorkflowRuntimeOptions.cs:34]: Param descriptions in function header (here and throughout)

File: examples/Workflow/WorkflowWebApp/Program.cs
Comment (tmacam) [reply examples/Workflow/WorkflowWebApp/Program.cs:50]: Sorry, misread your comment. Yes, changing from 8080.

File: examples/Workflow/WorkflowWebApp/WorkflowWebApp.csproj
Comment (tmacam) [reply examples/Workflow/WorkflowWebApp/WorkflowWebApp.csproj]: done

File: examples/Workflow/WorkflowWebApp/Program.cs
Comment (tmacam) [reply examples/Workflow/WorkflowWebApp/Program.cs]: Just removed this temporary variable that was being overwritten in the next line anyway.

File: src/Dapr.Workflow/WorkflowMetadata.cs
Comment (tmacam) [reply src/Dapr.Workflow/WorkflowMetadata.cs:23]: Metadata is a common idiom on Dapr, even on the .Net SDK. While I don't have strong opinions about this, I just want to make sure this is to align with .Net SDKs expectations or more a matter of preference.

Comment (cgillum) [reply src/Dapr.Workflow/WorkflowMetadata.cs:23]: I'm fine either way on this too. If "metadata" is too overloaded, then I think "details" is fine.

File: src/Dapr.Workflow/WorkflowContext.cs
Comment (cgillum) [reply src/Dapr.Workflow/WorkflowContext.cs:67]: No, cancellation tokens would make it easy for developers to do some wrong things with workflows (for example, you wouldn't want to cancel a durable workflow operation simply because the app host is shutting down), so we opted for a design that instead incorporated timeouts via TimeSpan.

File: src/Dapr.Workflow/WorkflowContext.cs
Comment (cgillum) [reply src/Dapr.Workflow/WorkflowContext.cs:75]: The base SDK doesn't support this, currently.

File: src/Dapr.Workflow/WorkflowMetadata.cs
Comment (halspang) [reply src/Dapr.Workflow/WorkflowMetadata.cs:23]: Metadata and Options are both used in the SDK. I'm fine either way, but would at least prefer the member to match as it feels weird to me to have something like `workflowMetadata.Details` and would prefer to either have `workflowDetails` or `workflowMetadata.OrchestrationMetadata`.

File: src/Dapr.Workflow/WorkflowRuntimeOptions.cs
Comment (halspang) [inline src/Dapr.Workflow/WorkflowRuntimeOptions.cs:66]: This adds Workflows and Activities, yes? I think I may be confused as to what the difference of those is or their relation to one another.

File: src/Dapr.Workflow/WorkflowServiceCollectionExtensions.cs
Comment (halspang) [reply src/Dapr.Workflow/WorkflowServiceCollectionExtensions.cs:31]: I'm fine either way. I'm not super familiar with the existing options though, so my concern would be if this would clobber anything. Do we expect people to have our libraries and the other workflow libraries together?

File: src/Dapr.Workflow/WorkflowServiceCollectionExtensions.cs
Comment (halspang) [inline src/Dapr.Workflow/WorkflowServiceCollectionExtensions.cs:40]: Do we think it's worth checking for this to have been registered already? Things like `DaprClient` do this:

```csharp
if (services.Any(s => s.ImplementationType == typeof(DaprClientMarkerService)))
{
    return;
}

services.AddSingleton<DaprClientMarkerService>();
```

**Review by halspang — COMMENTED** [review]
Mostly minor comments, honestly. Most comments stem from wanting to understand the workflow experience a bit more.

File: examples/Workflow/WorkflowWebApp/Program.cs
Comment (tmacam) [reply examples/Workflow/WorkflowWebApp/Program.cs:19]: I will let other correct me but if you think of a workflow as defining a graph, activities are the nodes. Both concepts are linked.  https://docs.temporal.io/activities

File: src/Dapr.Workflow/WorkflowContext.cs
Comment (cgillum) [reply src/Dapr.Workflow/WorkflowContext.cs:67]: Yes, I'm primarily worried that users might use it incorrectly since this is not a normal in-memory operation. It's very common in .NET to flow cancellation tokens to handle app shutdown events and it could result in unexpected behavior if developers blindly try to flow cancellation tokens to workflow context APIs. But I see your point that users may want to cancel for reasons other than timeout. I suggest we defer on this for now since the underlying API already doesn't support cancellation tokens (but I'm pinging the primary maintainer of this SDK to see if we should consider adding support for it later).

File: examples/Workflow/WorkflowWebApp/Program.cs
Comment (tmacam) [reply examples/Workflow/WorkflowWebApp/Program.cs:27]: Just to shorten it. Not required, the whole GUID could be used instead.

File: examples/Workflow/WorkflowWebApp/Program.cs
Comment (tmacam) [reply examples/Workflow/WorkflowWebApp/Program.cs:35]: I think that this example, as a web application might be making things a bit more complicated than they need to be.

It does make it simpler to handle the async nature of workflow execution by delegating the "status checking" to the user running the example. I think this should rather be a CLI that keeps pooling for a status change (just to make it more... interactive 😅), prints things as it goes and is done with it.

Thoughts?

File: examples/Workflow/WorkflowWebApp/Properties/launchSettings.json
Comment (tmacam) [reply examples/Workflow/WorkflowWebApp/Properties/launchSettings.json:1]: Ok. So we are keeping this. I will resolve this conversation then.

File: src/Dapr.Workflow/WorkflowContext.cs
Comment (cgillum) [reply src/Dapr.Workflow/WorkflowContext.cs:67]: Ugh, sorry, I was just corrected. We _do_ have an overload in the Durable Task SDK that accepts a cancellation token. The `TimeSpan` overload was mainly added for convenience. Since there's precedence for this, we can add a `WaitForExternalEventAsync` overload that takes a cancellation token if that would be preferred.

Comment (tmacam) [reply src/Dapr.Workflow/WorkflowContext.cs:67]: Do you folks think it would be worth also opening an issue on this SDK to keep this registered and so we revisit this in the future? @cgillum, is there a GitHub issue to link here or in this possible issue?

File: examples/Workflow/WorkflowWebApp/Program.cs
Comment (cgillum) [reply examples/Workflow/WorkflowWebApp/Program.cs:35]: This example is confusing because it treats workflows as first-class things, whereas a real app doesn't think in terms of workflows, but rather in terms of "orders", "jobs", etc. Ideally, this route would be something like "/orders/{orderId}" where the implementation maps the orderId to a workflow ID.

File: examples/Workflow/WorkflowWebApp/WorkflowWebApp.csproj
Comment (cgillum) [reply examples/Workflow/WorkflowWebApp/WorkflowWebApp.csproj:9]: As you mentioned, .NET 3.1 is no longer supported, so I figured it didn't make sense to add new code that requires .NET 3.1. However, the dependencies are all .NET Standard 2.0, so it's fine if you'd rather make this `netcore3.1` and then migrate everything wholesale later.

File: src/Dapr.Workflow/WorkflowContext.cs
Comment (cgillum) [reply src/Dapr.Workflow/WorkflowContext.cs:67]: Not sure if our replies raced with each other, but the Durable Task SDK already has support for cancellation tokens here. It's up to use if we want to expose it. We can either do that now, open an issue to add support later, or just wait for customer feedback to see if they actually need/want this.

File: examples/Workflow/WorkflowWebApp/Program.cs
Comment (tmacam) [reply examples/Workflow/WorkflowWebApp/Program.cs:35]: I did a small renaming inside this file which should be sent along the next commit.

File: examples/Workflow/WorkflowWebApp/Program.cs
Comment (tmacam) [reply examples/Workflow/WorkflowWebApp/Program.cs:48]: Updated the URL in the last commit.

File: src/Dapr.Workflow/WorkflowServiceCollectionExtensions.cs
Comment (tmacam) [reply src/Dapr.Workflow/WorkflowServiceCollectionExtensions.cs:40]: Curious: should this be done for each `AddSingleton<T>` ?

File: examples/Workflow/WorkflowWebApp/WorkflowWebApp.csproj
Comment (tmacam) [reply examples/Workflow/WorkflowWebApp/WorkflowWebApp.csproj:9]: The current example code as it is requires net6 :/

File: src/Dapr.Workflow/WorkflowServiceCollectionExtensions.cs
Comment (tmacam) [reply src/Dapr.Workflow/WorkflowServiceCollectionExtensions.cs]: Added an extension method to handle it

File: src/Dapr.Workflow/WorkflowRuntimeOptions.cs
Comment (tmacam) [reply src/Dapr.Workflow/WorkflowRuntimeOptions.cs:70]: You have a good point. @cgillum, is this the appropriate name for this?

File: src/Dapr.Workflow/WorkflowServiceCollectionExtensions.cs
Comment (tmacam) [inline src/Dapr.Workflow/WorkflowServiceCollectionExtensions.cs:63]: I wonder if I should move this to `DaprServiceCollectionExtensions` and refactor `AddDaprClient` to use it...

File: src/Dapr.Workflow/WorkflowMetadata.cs
Comment (tmacam) [reply src/Dapr.Workflow/WorkflowMetadata.cs:23]: Marking this as resolved

File: src/Dapr.Workflow/WorkflowContext.cs
Comment (tmacam) [reply src/Dapr.Workflow/WorkflowContext.cs:69]: I am exposing it in ad7ade292b2a495500da2f151b0f448e3c31fa53. Resolving this for now. Reopen if required.

File: src/Dapr.Workflow/WorkflowClient.cs
Comment (tmacam) [reply src/Dapr.Workflow/WorkflowClient.cs:53]: Done.

File: src/Dapr.Workflow/WorkflowRuntimeOptions.cs
Comment (tmacam) [reply src/Dapr.Workflow/WorkflowRuntimeOptions.cs:36]: Done. I did a quick check but let me know if anything is missing.

File: src/Dapr.Workflow/WorkflowRuntimeOptions.cs
Comment (cgillum) [inline src/Dapr.Workflow/WorkflowRuntimeOptions.cs:71]: It seems to me that the code in this method works for both workflows and activities. If we want to be really specific, we'd probably want to name this `AddWorkflowsAndActivitiesToRegistry`.

**Review by cgillum — APPROVED** [review]
I'm good with this PR as it is. There are a several additions I'd like to make, but those can be done in subsequent PRs.

File: src/Dapr.Workflow/WorkflowRuntimeOptions.cs
Comment (tmacam) [reply src/Dapr.Workflow/WorkflowRuntimeOptions.cs:71]: But it doesn't really add/register any workflow with the registry, does it?

File: src/Dapr.Workflow/WorkflowServiceCollectionExtensions.cs
Comment (halspang) [reply src/Dapr.Workflow/WorkflowServiceCollectionExtensions.cs]: I think we could refactor it in a future PR :)

File: examples/Workflow/WorkflowWebApp/Program.cs
Comment (halspang) [reply examples/Workflow/WorkflowWebApp/Program.cs]: I wouldn't say this is required, just musing since this is a preview/alpha feature:

That was my initial thought, but I guess my question was more do the activities need to be directly related to the workflow, or can they exist outside of one? Should we have some kind of validation that checks that "SayHello" is defiend? Or should we define the activity in the workflow registration/via actual object references?

File: src/Dapr.Workflow/WorkflowServiceCollectionExtensions.cs
Comment (tmacam) [reply src/Dapr.Workflow/WorkflowServiceCollectionExtensions.cs]: I talked to @philliphoff  and I am using TryAddSingleton which does exactly the same verifications and addition and is part of [`microsoft.extensions.dependencyinjection.extensions`](https://learn.microsoft.com/en-us/dotnet/api/microsoft.extensions.dependencyinjection.extensions.servicecollectiondescriptorextensions.tryaddsingleton?view=dotnet-plat-ext-7.0#microsoft-extensions-dependencyinjection-extensions-servicecollectiondescriptorextensions-tryaddsingleton-1(microsoft-extensions-dependencyinjection-iservicecollection)).

File: examples/Workflow/WorkflowWebApp/Program.cs
Comment (halspang) [reply examples/Workflow/WorkflowWebApp/Program.cs:50]: You should be able to remove the URL from `app.Run()` then, yes?

File: src/Dapr.Workflow/WorkflowRuntimeOptions.cs
Comment (halspang) [reply src/Dapr.Workflow/WorkflowRuntimeOptions.cs:71]: I'm in favor of being more specific here, though if anyone feels super strongly we can leave it as is.

**Review by halspang — COMMENTED** [review]
Overall, the code seems fine. I think the big thing for me is that we should remove the example entirely to start off. Examples should show how we want users to invoke/use the code instead of just providing the syntax. 

What I'd prefer to do here is create an issue to create a workflow and once the code is complete, spend some time to make a more real-world example that shows workflows as part of a larger picture.

File: src/Dapr.Workflow/WorkflowRuntimeOptions.cs
Comment (tmacam) [reply src/Dapr.Workflow/WorkflowRuntimeOptions.cs]: Done in 509656b35a9282d4740a781c1800005b96b36262

Comment (tmacam) [reply src/Dapr.Workflow/WorkflowRuntimeOptions.cs]: Resolving.

File: src/Dapr.Workflow/WorkflowRuntimeOptions.cs
Comment (tmacam) [reply src/Dapr.Workflow/WorkflowRuntimeOptions.cs]: Renamed. Resolving.

File: src/Dapr.Workflow/WorkflowServiceCollectionExtensions.cs
Comment (tmacam) [reply src/Dapr.Workflow/WorkflowServiceCollectionExtensions.cs]: And this was addressed using `TryAddSingleton`. Resolving.

**Comment by tmacam** [comment]
> Overall, the code seems fine. I think the big thing for me is that we should remove the example entirely to start off. Examples should show how we want users to invoke/use the code instead of just providing the syntax.
> 
> What I'd prefer to do here is create an issue to create a workflow and once the code is complete, spend some time to make a more real-world example that shows workflows as part of a larger picture.

@cgillum, you will be updating the example in a follow up PR, yes? So I will defer comment regarding the removal of the example to you.

**Comment by cgillum** [comment]
Yes, I can update the example to be closer to the recent demo I gave. I agree that would be much better. However, I don't necessarily agree that we should exclude the existing example from this app. That's just creating unnecessary work for us, and I think it's honestly better than having nothing at all.

File: examples/Workflow/WorkflowWebApp/Program.cs
Comment (tmacam) [reply examples/Workflow/WorkflowWebApp/Program.cs]: Fixed in fb8589cf6100f7481532685e77de911e3ca71954. Resolving.

**Comment by halspang** [comment]
> Yes, I can update the example to be closer to the recent demo I gave. I agree that would be much better. However, I don't necessarily agree that we should exclude the existing example from this PR. That's just creating unnecessary work for us, and I think it's honestly better than having nothing at all.

If it we're planning on updating it shortly, that works for me. I just wanted to make sure we had a full example before we eventually hit a full release for workflows.

**Review by halspang — APPROVED** [review]

**Pull request merged**
